Assets/Scripts/DialogueManager.cs

```diff
@@ -219,72 +261,90 @@ private IEnumerator StartTypeText(string text)
             textMesh.text = text.Substring(0, currentCharIndex);
             yield return new WaitForSeconds(charTypeSpeed);
         }
-
-        waitingForNext = true;
-        while (waitingForNext)
+        activeConversations[conversation].waitingForNext = true;
+        while (activeConversations[conversation].waitingForNext)
         {
             yield return new WaitForFixedUpdate();
         }
 
         textMesh.text = string.Empty;
 
-        skipRequest = false;
-        isTyping = false;
+        activeConversations[conversation].skipRequest = false;
+        activeConversations[conversation].isTyping = false;
+
+        AdvanceConversation(conversation);
+    }
+
+    private void FinishConversation(Conversation conversation)
+    {
+        HideAllSpeakers(conversation);
+
+        // Stop any coroutines that are part of the conversation
+        // Null checks are done to be super safe just incase coroutines have not been set
+        if (activeConversations[conversation].coroutine != null)
+        {
+            StopCoroutine(activeConversations[conversation].coroutine);
+        }
+        if (activeConversations[conversation].typingCoroutine != null)
+        {
+            StopCoroutine(activeConversations[conversation].typingCoroutine);
+        }
+
+        activeConversations.Remove(conversation);
+
+        if (advancingConversation == conversation)
+        {
+            EndAdvancing();
+        }
+    }
+
+    void HideAllSpeakers(Conversation conversation)
+    {
+        List<string> convoSpeakers = conversation.GetAllSpeakers();
 
-        AdvanceConversation();
+        foreach(string speaker in convoSpeakers)
+        {
+            SpeakerUI speakerUI = Utils.GetRequiredComponent<SpeakerUI>(speakers[speaker].speakerObject);
```

Comment:
This getting of component seems common... maybe we should get this once per speaker so we limit how many times we need to do GetComponent on them

Reply:
It is common and maybe it's just me being lazy but it's much easier just getting the SpeakerUI from the actual speaker object. If we were to store the speakers then we would have to do additional checks to find the speakers within the list and then add/remove them also. If you think that's better then I don't mind changing it.

Reply:
Was just thinking of having the SpeakerUI be another thing in the SpeakerContainer (that isn't set in the inspector though and is just automatically taken from the GameObject.).
Maybe when initializing the dictionary you can instead call AddSpeaker function and in that function you can do the one-time GetComponent call? Would make it so adding a speaker from inspector and adding a speaker at runtime go through the same logic.